### PR TITLE
Collect usage data from files in the Log folder and send them to a database

### LINKED ISSUE #1: Collect usage data from files in the Log folder and send them to a database
#### Description:
Initial version of our usage Analytics. The targeted database should be provided in the setting folders from ProgramData BHoM

### PR DESCRIPTION
**Pull request by adecler**
### NOTE: Depends on 
https://github.com/BHoM/BHoM_UI/pull/295

   
### Issues addressed by this PR

Closes #1

There's two sections to the new toolkit:
- The engine is in charge of collecting the usage data from the log files (and delete them afterwards if requested). as any other engine method, it is available in the UI. Be aware that it will never read the data from the current log, only from previous log files.
- The adapter isn't really a typical adapter at the moment as it doesn't implement any push or pull method. It is purely there to provide the capability to send the usage data to the database when the UI starts. In the long run, we might start adding pulling functionality to get analytical information back.

#### To test that the Engine is correctly collecting data:

image

The default value for deleting the processed files false so don't worry :-)

Note that you might have to recompile BHoM_UI and Grasshopper since it looks like the code that creates the log files is not in the installer yet.

if you want to check your log files, you can find them in `C:\ProgramData\BHoM\Logs`

#### To test sending the data to a database:

image

Once you have done that, the toolkit should start sending data to the database at every start of the UI. Be aware that it will delete the processed log files.
Obviously, you might want to switch to the London server or you localhost instead 😄

### CODE REVIEW

**Comment by adecler** [comment]
Getting there 😄 

image

Just need to add some multi-threading to make sure it doesn't stall the whole thing when you create you first component

**Comment by adecler** [comment]
This one is ready for review too

File: BHoMAnalytics_Engine/BHoMAnalytics_Engine.csproj [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    <Reference Include="System" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml.Linq" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Net.Http" />
+    <Reference Include="System.Xml" />
+    <Reference Include="UI_oM">
+      <HintPath>.C:\ProgramData\BHoM\Assemblies\UI_oM.dll</HintPath>
```
Comment (al-fisher) [inline BHoMAnalytics_Engine/BHoMAnalytics_Engine.csproj:55]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\UI_oM.dll</HintPath>
```

**Comment by al-fisher** [comment]
Can confirm this is working for me - with expected behaviour of recording logs, sending to (my locally hosted) database, and removing logs on successful execution.
image

Looking good.

I did have some odd behaviour on set up. The SaveSettings Method (seemed to) fail to successfully write to the file - until I manually created a version. It was then able to update.

I could however not recreate this behaviour 😕 sorry. So, as will be needed anyway, good field testing on more users machines will be important - but can be post merge I would say, so we can unblock testing hosting etc.

The SaveSettings etc, will be carried out as part of the installer in any case - and so that whole process will be tested at large too

**Comment by al-fisher** [comment]
Small typo above - other than that - think we are good to merge ready for wider testing! 
Cheers @adecler

**Pull request merged**